File: src/receiver.rs
/****************************************************************************
 * Copyright (C) 2020 Pier Luigi Fiorini <pierluigi.fiorini@gmail.com>
 *
 * SPDX-License-Identifier: GPL-3.0-or-later
 ***************************************************************************/

use crate::api;
use crate::errors::GenericError;
use log::info;
use ostree;
use std::collections::HashMap;
use std::fs;
use std::path::Path;
use std::path::PathBuf;

pub struct Receiver {
    repo_path: PathBuf,
    temp_path: PathBuf,
}

impl Receiver {
    pub fn new(repo_path: &Path) -> Result<Receiver, GenericError> {
        // Create temporary directory
        let temp_path = repo_path.join(".tmp");
        fs::create_dir_all(&temp_path).map_err(|e| {
            GenericError::new(&format!("Failed to create temporary directory: {}", e))
        })?;

        Ok(Receiver {
            repo_path: repo_path.to_owned(),
            temp_path: temp_path.to_owned(),
        })
    }

    fn open_repo(&self) -> Result<ostree::Repo, GenericError> {
        // We can't save an ostree::Repo instance inside Receiver because
        // it cannot be assigned to threads, so we reopen it every time
        // we need it
        let cancellable = gio::Cancellable::new();
        let repo = ostree::Repo::new_for_path(&self.repo_path);
        repo.open(Some(&cancellable))
            .map_err(|e| GenericError::new(&format!("Failed to open the repository: {}", e)))?;
        Ok(repo)
    }

    pub fn temp_path(&self, filename: &str) -> PathBuf {
        self.temp_path.join(&filename)
    }

    pub fn obj_path(&self, filename: &str) -> PathBuf {
        self.repo_path
            .join("objects")
            .join(&filename[..2])
            .join(&filename[2..])
    }

    pub fn get_info(&self) -> Result<api::Info, GenericError> {
        let repo = self.open_repo()?;

        let mode = match repo.get_mode() {
            ostree::RepoMode::Bare => "bare",
            ostree::RepoMode::Archive => "archive",
            ostree::RepoMode::BareUser => "bare-user",
            ostree::RepoMode::BareUserOnly => "bare-user-only",
            _ => "unknown",
        };

        let cancellable = gio::Cancellable::new();
        let refs = repo
            .list_refs(None, Some(&cancellable))
            .map_err(|e| GenericError::new(&format!("Failed to list refs: {}", e)))?;

        Ok(api::Info {
            mode: String::from(mode),
            refs: refs,
        })
    }

    pub fn check_update(
        &self,
        refs: HashMap<String, (String, String)>,
    ) -> Result<api::Status, GenericError> {
        let repo = self.open_repo()?;

        for (branch, revs) in refs {
            // See if branch can be updated (pass allow_noent=false otherwise it will
            // crash when the branch doesn't exist)
            match repo.resolve_rev(&branch, false) {
                Err(_) => {
                    // Branch cannot be resolved on the client end
                    if revs.0 != api::REV_NULL {
                        return Ok(api::Status {
                            status: false,
                            message: Some(format!(
                                "Invalid from commit {} for new branch {}",
                                &revs.0, &branch
                            )),
                        });
                    }
                }
                Ok(current) => {
                    if revs.0 != current {
                        return Ok(api::Status {
                            status: false,
                            message: Some(format!(
                                "Branch {} is at {}, not {}",
                                &branch, &current, &revs.0
                            )),
                        });
                    }
                }
            }
        }

        Ok(api::Status {
            status: true,
            message: None,
        })
    }

    pub fn update_refs(&self, refs: HashMap<String, (String, String)>) -> Result<(), GenericError> {
        let repo = self.open_repo()?;
        let cancellable = gio::Cancellable::new();

        for (branch, revs) in refs {
            info!(
                "Setting branch {} revision from {} to {}",
                &branch, &revs.0, &revs.1
            );
            repo.set_ref_immediate(None, &branch, Some(&revs.1), Some(&cancellable))
                .map_err(|e| {
                    GenericError::new(&format!(
                        "Failed to set branch {} revision from {} to {}: {}",
                        &branch, &revs.0, &revs.1, e
                    ))
                })?;
        }

        Ok(())
    }
}

File: internal/receiver/receiver.go
// SPDX-FileCopyrightText: 2020 Pier Luigi Fiorini <pierluigi.fiorini@gmail.com>
//
// SPDX-License-Identifier: AGPL-3.0-or-later

package receiver

import (
	"fmt"
	"os"
	"path/filepath"

	"github.com/lirios/ostree-upload/internal/common"
	"github.com/lirios/ostree-upload/internal/ostree"
)

// ContextKey is a type that represent the key of a context
type ContextKey int

const (
	// KeyQueue is the context key for the update queue
	KeyQueue ContextKey = iota

	// KeyRepository is the context key for the ostree.Repo instance
	KeyRepository ContextKey = iota
)

// Name of the temporary directory inside the OSTree repository
const tempDirName = "tmp/ostree-upload"

// CreateTempDirectory creates a temporary directory inside the repository, used to store the objects during the upload
func CreateTempDirectory(r *ostree.Repo) error {
	tempPath := filepath.Join(r.Path(), tempDirName)

	// Check if the temporary directory already exist
	if _, err := os.Stat(tempPath); !os.IsNotExist(err) {
		return err
	}

	// Create temporary directory
	if err := os.Mkdir(tempPath, 0755); err != nil {
		return err
	}

	return nil
}

// GetTempObjectPath returns the path to the OSTree object passed as argument
// from the temporary directory
func GetTempObjectPath(r *ostree.Repo, objectName string) string {
	return filepath.Join(r.Path(), tempDirName, objectName)
}

// UpdateRefs points branches to the new checksum
func UpdateRefs(r *ostree.Repo, refs map[string]common.RevisionPair) error {
	for branch, revPair := range refs {
		if err := r.SetRefImmediate("", branch, revPair.Client); err != nil {
			return fmt.Errorf("Failed to set branch %s from %s to %s: %v", branch, revPair.Server, revPair.Client, err)
		}
	}

	return nil
}
